src/main/java/com/example/bookshop/service/impl/UserServiceImpl.java

```diff
@@ -0,0 +1,48 @@
+package com.example.bookshop.service.impl;
+
+import com.example.bookshop.dto.user.UserRegistrationRequestDto;
+import com.example.bookshop.dto.user.UserRegistrationResponseDto;
+import com.example.bookshop.exception.EntityNotFoundException;
+import com.example.bookshop.exception.RegistrationException;
+import com.example.bookshop.mapper.UserMapper;
+import com.example.bookshop.model.Role;
+import com.example.bookshop.model.User;
+import com.example.bookshop.repository.role.RoleRepository;
+import com.example.bookshop.repository.user.UserRepository;
+import com.example.bookshop.service.UserService;
+import java.util.Set;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleRepository roleRepository;
+    private Role role;
+
+    public UserServiceImpl(UserRepository userRepository, PasswordEncoder passwordEncoder,
+                           UserMapper userMapper, RoleRepository roleRepository) {
+        this.userRepository = userRepository;
+        this.passwordEncoder = passwordEncoder;
+        this.userMapper = userMapper;
+        this.roleRepository = roleRepository;
+    }
+
+    @Override
+    public UserRegistrationResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findUserByEmail(request.email()).isPresent()) {

```

Comment:
It's better to create boolean existsByEmail(String email) method in repo and use it here